hummingbot/connector/exchange/ascend_ex/ascend_ex_api_user_stream_data_source.py

```diff
@@ -63,84 +65,56 @@ async def listen_for_user_stream(self, ev_loop: asyncio.BaseEventLoop, output: a
-                                    "Unexpected error when parsing AscendEx message. ", exc_info=True
-                                )
-                                raise
-                    except Exception:
-                        self.logger().error(
-                            "Unexpected error while listening to AscendEx messages. ", exc_info=True
-                        )
-                        raise
-                    finally:
-                        await ws.close()
+                ws: WSAssistant = await self._get_ws_assistant()
+                url = f"{get_ws_url_private(accountGroup)}/{CONSTANTS.STREAM_PATH_URL}"
+                await ws.connect(ws_url=url, ws_headers=headers, ping_timeout=self.HEARTBEAT_PING_INTERVAL)
+
+                subscribe_request: WSRequest = WSRequest(payload)
+                async with self._throttler.execute_task(CONSTANTS.SUB_ENDPOINT_NAME):
+                    await ws.send(subscribe_request)
+
+                async for raw_msg in ws.iter_messages():
+                    msg = raw_msg.data
+                    if msg is None:
+                        continue
+                    self._last_recv_time = time.time()
+                    output.put_nowait(msg)
             except asyncio.CancelledError:
                 raise
             except Exception:
                 self.logger().error(
-                    "Unexpected error with AscendEx WebSocket connection. " "Retrying after 30 seconds...", exc_info=True
+                    "Unexpected error with AscendEx WebSocket connection. " "Retrying after 30 seconds...",
+                    exc_info=True
                 )
                 await asyncio.sleep(30.0)
-
-    async def _iter_messages(
-        self,
-        ws: aiohttp.ClientWebSocketResponse
-    ) -> AsyncIterable[str]:
-        # Terminate the recv() loop as soon as the next message timed out, so the outer loop can reconnect.
-        try:
-            while True:
-                raw_msg = await ws.receive()
-                if raw_msg.type == aiohttp.WSMsgType.CLOSED:
-                    raise ConnectionError
-                self._last_recv_time = time.time()
-                yield raw_msg.data
-
-        except asyncio.TimeoutError:
-            self.logger().warning("WebSocket ping timed out. Going to reconnect...")
-            return
-        finally:
-            await ws.close()
-
-    async def _handle_ping_message(self, ws: aiohttp.ClientWebSocketResponse):
-        async with self._throttler.execute_task(CONSTANTS.PONG_ENDPOINT_NAME):
-            pong_payload = {
-                "op": "pong"
-            }
-            await ws.send_json(pong_payload)
+            finally:
+                ws and await ws.disconnect()
+
+    async def _get_rest_assistant(self) -> RESTAssistant:
+        if self._rest_assistant is None:
+            self._rest_assistant = await self._api_factory.get_rest_assistant()
+        return self._rest_assistant
```

Comment:
This is only used in one place. There is no need to have the function and the _rest_assitant variable

Reply:
I put it there for consistency with other classes.
Should I remove it then?

Reply:
I would say we should remove it if not required. We need to change the other connectors when the same happens. Are you referring to Binance connectors?

Reply:
I'm referring to the _ws_assistant and _get_ws_assistant() , which is required in user stream to get the last_recv_time.

Reply:
Removed.